// ngx-apexcharts/esm2022/lib/ngx-apexcharts.module.mjs
import { NgModule } from '@angular/core';
import { ChartComponent } from './chart/chart.component';
import * as i0 from "@angular/core";
export class NgxApexchartsModule {
    /** @nocollapse */ static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: NgxApexchartsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    /** @nocollapse */ static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.0.4", ngImport: i0, type: NgxApexchartsModule, imports: [ChartComponent], exports: [ChartComponent] }); }
    /** @nocollapse */ static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: NgxApexchartsModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: NgxApexchartsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [],
                    imports: [ChartComponent],
                    exports: [ChartComponent],
                }]
        }] });
